### PR TITLE
GH-34968: [C++] Add Equal Options to RecordBatch

### LINKED ISSUE #34968: [C++] RecordBatch support EqualOptions in `Equals` and `ApproxEquals`
### Describe the enhancement requested

Currently, directly comparing `Array` supports `EqualOptions`, which support nan comparing mode and others. However, array doesn't support passing argument like that. So, it can only use default `EqualOptions`

### Component(s)

C++

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

Currently, directly comparing Array supports EqualOptions, which support nan comparing mode and others. However, array doesn't support passing argument like that. So, it can only use default EqualOptions

### What changes are included in this PR?

Add EqualOptions 

### Are these changes tested?

currently not

### Are there any user-facing changes?

User can call EqualOptions in RecordBatch

* Closes: #34968

### CODE REVIEW

**Comment by kou** [comment]
Could you fix a CI failure?

https://github.com/apache/arrow/actions/runs/4639277857/jobs/8209947441?pr=34970#step:6:3002

```text
/arrow/cpp/src/arrow/record_batch.h:107: error: The following parameter of arrow::RecordBatch::ApproxEquals(const RecordBatch &other, const EqualOptions &opts=EqualOptions::Defaults()) const is not documented:
  parameter 'other' (warning treated as error, aborting now)
```

**Comment by mapleFU** [comment]
Sure, trying to fix it and add test now

File: cpp/src/arrow/record_batch_test.cc [hunk 1]
```diff
@@ -81,6 +82,40 @@ TEST_F(TestRecordBatch, Equals) {
...
+
+    auto b1 = RecordBatch::Make(schema, length, {array1});
+    auto b2 = RecordBatch::Make(schema, length, {array2});
+
+    EXPECT_FALSE(b1->ApproxEquals(*b2, EqualOptions::Defaults().nans_equal(false)));
+    EXPECT_FALSE(b1->ApproxEquals(*b2, EqualOptions::Defaults().nans_equal(true)));
+
+    EXPECT_TRUE(
+        b1->ApproxEquals(*b2, EqualOptions::Defaults().nans_equal(true).atol(0.1)));
+  }
```
Comment (kou) [inline cpp/src/arrow/record_batch_test.cc:116]: Could you extract this to a separated test?

File: cpp/src/arrow/record_batch.h [hunk 2]
```diff
@@ -97,11 +98,17 @@ class ARROW_EXPORT RecordBatch {
...
   /// \return true if batches are equal
-  bool Equals(const RecordBatch& other, bool check_metadata = false) const;
+  bool Equals(const RecordBatch& other, bool check_metadata = false,
+              const EqualOptions& opts = EqualOptions::Defaults()) const;
 
   /// \brief Determine if two record batches are approximately equal
-  bool ApproxEquals(const RecordBatch& other) const;
+  ///
+  /// \param[in] other the RecordBatch to compare with
+  /// \param[in] opts the options for equality comparisons
```
Comment (wgtmac) [inline cpp/src/arrow/record_batch.h:109]: nit: add `\return` as well

File: cpp/src/arrow/record_batch_test.cc (hunk 1 above)
Comment (mapleFU) [reply cpp/src/arrow/record_batch_test.cc]: Sure, done

**Comment by mapleFU** [comment]
@kou I've fixed all comments. But `continuous-integration/appveyor/pr` failed, and I don't know why it failed, would you mind take a look?

File: cpp/src/arrow/record_batch.h [hunk 3]
```diff
@@ -97,11 +98,18 @@ class ARROW_EXPORT RecordBatch {
...
+  /// \return true if batches are equal
```
Comment (westonpace) [inline cpp/src/arrow/record_batch.h:110]:
```suggestion
  /// \return true if batches are approximately equal
```

File: cpp/src/arrow/record_batch_test.cc [hunk 4]
```diff
@@ -81,6 +82,46 @@ TEST_F(TestRecordBatch, Equals) {
   ASSERT_FALSE(b1->Equals(*b2, /*check_metadata=*/true));
 }
 
+TEST_F(TestRecordBatch, EqualsOptions) {
```
Comment (westonpace) [inline cpp/src/arrow/record_batch_test.cc:85]:
```suggestion
TEST_F(TestRecordBatch, EqualOptions) {
```

File: cpp/src/arrow/record_batch_test.cc [hunk 5]
```diff
@@ -81,6 +82,46 @@ TEST_F(TestRecordBatch, Equals) {
...
+  int length = 2;
+  auto f = field("f", float64());
+
+  auto metadata = key_value_metadata({"foo"}, {"bar"});
+
```
Comment (westonpace) [inline cpp/src/arrow/record_batch_test.cc:90]:
```suggestion
```

File: cpp/src/arrow/record_batch_test.cc [hunk 6]
```diff
@@ -81,6 +82,46 @@ TEST_F(TestRecordBatch, Equals) {
...
+  auto schema = ::arrow::schema(fields);
+
+  std::shared_ptr<Array> array1, array2;
+  ArrayFromVector<DoubleType>(float64(), {true, true}, {0.5, NAN}, &array1);
+  ArrayFromVector<DoubleType>(float64(), {true, true}, {0.5, NAN}, &array2);
+  auto b1 = RecordBatch::Make(schema, length, {array1});
+  auto b2 = RecordBatch::Make(schema, length, {array2});
+
+  EXPECT_FALSE(b1->Equals(*b2, false, EqualOptions::Defaults().nans_equal(false)));
+  EXPECT_TRUE(b1->Equals(*b2, false, EqualOptions::Defaults().nans_equal(true)));
```
Comment (westonpace) [inline cpp/src/arrow/record_batch_test.cc:101]:
```suggestion
  EXPECT_FALSE(b1->Equals(*b2, /*check_metadata=*/ false, EqualOptions::Defaults().nans_equal(false)));
  EXPECT_TRUE(b1->Equals(*b2, /*check_metadata=*/ false, EqualOptions::Defaults().nans_equal(true)));
```

File: cpp/src/arrow/record_batch_test.cc [hunk 7]
```diff
@@ -81,6 +82,46 @@ TEST_F(TestRecordBatch, Equals) {
...
+}
+
+TEST_F(TestRecordBatch, ApproxEqualsOptions) {
+  int length = 2;
+  auto f = field("f", float64());
+
+  auto metadata = key_value_metadata({"foo"}, {"bar"});
+
```
Comment (westonpace) [inline cpp/src/arrow/record_batch_test.cc:109]:
```suggestion
```

**Review by westonpace — CHANGES_REQUESTED** [review]
Very minor nits.  I like this change and it looks good.  Thanks!

The Appveyor failure is unrelated.

**Review by westonpace — APPROVED** [review]
Thanks for the quick turnaround

**Pull request merged**